### PR TITLE
Chore end to end tests for multiple golem messages routing

### PR DESCRIPTION
**Pull request by Jakub89**
Resolves: #925

### CODE REVIEW

File: concent_api/api_testing_common.py [hunk 1]
```diff
@@ -153,15 +152,37 @@ def validate_content_type(actual_content_type: str, expected_content_type: Optio
         )
 
 
+def validate_golem_version(actual_golem_version: str, expected_golem_version: Optional[str]=None) -> None:
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: Why is `expected_golem_version` optional? Is there any case where it will be `None`? If not, please make it non-optional.

File: concent_api/api_testing_common.py [hunk 2]
```diff
@@ -153,15 +152,37 @@ def validate_content_type(actual_content_type: str, expected_content_type: Optio
...
+    if expected_golem_version is not None:
+        assert_condition(
+            '.'.join(actual_golem_version.split('.')[:-1]),
+            '.'.join(expected_golem_version.split('.')[:-1]),
+            f"Expected Golem Message version: {'.'.join(expected_golem_version.split('.')[:-1])}.X, "
+            f"actual version: {actual_golem_version}"
+        )
+
+
+def validate_error_code(actual_error_code: str, expected_error_code: Optional[str]=None) -> None:
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: Same here.

File: concent_api/api_testing_common.py [hunk 3]
```diff
@@ -153,15 +152,37 @@ def validate_content_type(actual_content_type: str, expected_content_type: Optio
...
+    if expected_error_code is not None:
+        assert_condition(
+            actual_error_code,
+            expected_error_code,
+            f"Wrong content type for Golem Message: {actual_error_code}"
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: I think this message is not related with this function purpose.

File: concent_api/api_testing_common.py [hunk 4]
```diff
@@ -185,21 +206,24 @@ def _print_data(data: Optional[Union[bytes, str]], url: str) -> None:
...
     url = "{}/api/v1/{}/".format(host, endpoint)
-
     _print_data(data, url)
-    response = requests.post("{}".format(url), headers=REQUEST_HEADERS, data=_prepare_data(data), verify=False)
+    response = requests.post("{}".format(url), headers=headers, data=_prepare_data(data), verify=False)
     _print_response(private_key, public_key, response)
     validate_response_status(response.status_code, expected_status)
     validate_content_type(response.headers['Content-Type'], expected_content_type)
     validate_response_message(response.content, expected_message_type, private_key, public_key)
+    if 'Concent-Golem-Messages-Version' in response.headers:
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: Maybe we should we print something when this header is missing?
And I see now below that `expected_golem_version` can be None with this flow. I think it would be better to add to this `if `:
```
if expected_golem_version is not None and 'Concent-Golem-Messages-Version' in response.headers:
```
and the make it non-optional in function.

File: concent_api/api_testing_common.py [hunk 5]
```diff
@@ -185,21 +206,24 @@ def _print_data(data: Optional[Union[bytes, str]], url: str) -> None:
...
+        validate_golem_version(response.headers['Concent-Golem-Messages-Version'], expected_golem_version)
     print()
     content_type = response.headers['Content-Type']
     if 'text/html' in content_type:
         return None
-    elif content_type == 'json':
+    elif content_type == 'application/json':
+        validate_error_code((json.loads(response.content))['error_code'], expected_error_code)
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: Same here.

File: concent_api/api_testing_common.py [hunk 6]
```diff
@@ -268,23 +292,24 @@ def create_client_auth_message(client_priv_key: bytes, client_public_key: bytes,
     return dump(client_auth, client_priv_key, concent_public_key)
 
 
-def parse_command_line(command_line: list) -> str:
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: :+1: for refactoring this.

File: concent_api/api_testing_common.py [hunk 7]
```diff
@@ -268,23 +292,24 @@ def create_client_auth_message(client_priv_key: bytes, client_public_key: bytes,
...
-    return cluster_url
-
-
 def parse_arguments() -> Tuple:
     parser = argparse.ArgumentParser()
     parser.add_argument("cluster_url")
     parser.add_argument("tc_patterns", nargs='*')
+    parser.add_argument(
+        '-c1',
+        '--concent_1_version',
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: I think it should be named `--concent_1_golem_messages_version`. Same for second one.

File: concent_api/api_testing_common.py [hunk 8]
```diff
@@ -308,7 +333,10 @@ def execute_tests(tests_to_execute: list, objects: dict, **kwargs: Any) -> None:
 def run_tests(objects: dict, additional_arguments: Optional[dict]=None) -> None:
     if additional_arguments is None:
         additional_arguments = {}
-    (cluster_url, patterns) = parse_arguments()
+    (cluster_url, patterns, concent_1, concent_2) = parse_arguments()
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: Same here, longer names but more meaningful.

File: concent_api/api-e2e-cluster-multiple-golem-messages-routing.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+from golem_messages.message import Ping
+
+from api_testing_common import api_request
+from api_testing_common import count_fails
+from api_testing_common import PROVIDER_PRIVATE_KEY
+from api_testing_common import run_tests
+from protocol_constants import ProtocolConstants
+
+import requests
```
Comment (rwrzesien) [inline concent_api/api-e2e-cluster-multiple-golem-messages-routing.py]: 3rd party lib should be before application specific imports (but pylint will not show it here as it is not checked by it).

File: concent_api/api_testing_common.py [hunk 10]
```diff
@@ -13,7 +13,6 @@
 import http.client
 import json
 import requests
-import sys
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: :+1:

File: concent_api/api-e2e-cluster-multiple-golem-messages-routing.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "concent_api.settings")
+
+
+@count_fails
+def test_case_ping_message_send_to_concent_with_supported_golem_messages_should_get_http400_response_with_message_unknown_error(
```
Comment (rwrzesien) [inline concent_api/api-e2e-cluster-multiple-golem-messages-routing.py]: Maybe we should write a short multi line comment (each line starting with #) why we expect 400 for supported version? It might be confusing for someone not really familiar with the case. The text can be copied from the issue (with little modification).

File: concent_api/api_testing_common.py [hunk 12]
```diff
@@ -308,7 +333,10 @@ def execute_tests(tests_to_execute: list, objects: dict, **kwargs: Any) -> None:
 def run_tests(objects: dict, additional_arguments: Optional[dict]=None) -> None:
     if additional_arguments is None:
         additional_arguments = {}
-    (cluster_url, patterns) = parse_arguments()
+    (cluster_url, patterns, concent_1, concent_2) = parse_arguments()
+    if concent_1 is not None and concent_2 is not None:
```
Comment (rwrzesien) [inline concent_api/api_testing_common.py]: I think there should be a check that golem-messages versions are not the same.

File: concent_api/api-e2e-cluster-multiple-golem-messages-routing.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        'send',
+        PROVIDER_PRIVATE_KEY,
+        CONCENT_PUBLIC_KEY,
+        Ping(),
+        headers={
+            'Content-Type': 'application/octet-stream',
+            'X-Golem-Messages': '1.0.0'
+        },
+        expected_status=404,
+        expected_error_code='not-found',
```
Comment (rwrzesien) [inline concent_api/api-e2e-cluster-multiple-golem-messages-routing.py]: Why this error code? I can't find in `common/constants.py`.

File: concent_api/api_testing_common.py (hunk 5 above)
Comment (Jakub89) [reply concent_api/api_testing_common.py]: Well yes, and it's the only time we can get `json` response from concent, so I thought it's all good?

File: concent_api/api_testing_common.py (hunk 12 above)
Comment (Jakub89) [reply concent_api/api_testing_common.py]: If they going to be the same, test will work just fine :P but I guess I could add that sanity check.

File: concent_api/api_testing_common.py (hunk 4 above)
Comment (Jakub89) [reply concent_api/api_testing_common.py]: This header is missing when we get http404 responses for example, so what do You suggest we should print about that?

File: concent_api/api_testing_common.py (hunk 7 above)
Comment (Jakub89) [reply concent_api/api_testing_common.py]: Done

File: concent_api/api_testing_common.py (hunk 8 above)
Comment (Jakub89) [reply concent_api/api_testing_common.py]: Done

File: concent_api/api-e2e-cluster-multiple-golem-messages-routing.py (hunk 9 above)
Comment (Jakub89) [reply concent_api/api-e2e-cluster-multiple-golem-messages-routing.py]: Done

File: concent_api/api-e2e-cluster-multiple-golem-messages-routing.py (hunk 11 above)
Comment (Jakub89) [reply concent_api/api-e2e-cluster-multiple-golem-messages-routing.py]: Added

File: concent_api/api_testing_common.py (hunk 12 above)
Comment (Jakub89) [reply concent_api/api_testing_common.py]: Done

File: concent_api/api_testing_common.py (hunk 4 above)
Comment (rwrzesien) [reply concent_api/api_testing_common.py]: You can now remove `if expected_golem_version is not None:` from the `validate_golem_version`.

Comment (Jakub89) [reply concent_api/api_testing_common.py]: Right! Done.

File: concent_api/api_testing_common.py (hunk 5 above)
Comment (rwrzesien) [reply concent_api/api_testing_common.py]: Please add `and expected_error_code is not None` to condition, that is what I have meant by 'same here'. Then please remove `if` from `validate_error_code` function.

About the error codes, I couldn't find a place when we set something else than our custom error codes in response, that is why I was asking about it. If you are sure there is something like `bad-response` in `error_code` field it would be good to know why ;) I currently don't know.

Comment (Jakub89) [reply concent_api/api_testing_common.py]: Now that You mention that, `bad-request` and `not-found` might be some default strings.
``` python
Running TC: test_case_ping_message_send_to_concent_with_malformed_golem_messages_should_get_http400_response_with_bad_request_error
SEND (http://devel.concent.golem.network/api/v1/send/)
MESSAGE:
    type                           = Ping
    timestamp                      = 1542196028
    encrypted                      = False
    sig                            = None
    header                         = MessageHeader(type_=1001, timestamp=1542196028, encrypted=False)
STATUS: 400 Bad Request
{'error_message': 'Bad Request', 'error_code': 'bad-request'}

Running TC: test_case_ping_message_send_to_concent_with_unsupported_golem_messages_should_get_http404_response_with_not_found_error
SEND (http://devel.concent.golem.network/api/v1/send/)
MESSAGE:
    type                           = Ping
    timestamp                      = 1542196028
    encrypted                      = False
    sig                            = None
    header                         = MessageHeader(type_=1001, timestamp=1542196028, encrypted=False)
STATUS: 404 Not Found
{'error_message': 'Not Found', 'error_code': 'not-found'}
```

**Comment by Jakub89** [comment]
@rwrzesien It looks like `bad-request` and `not-found` are sent by `nginx` that's why we can't find that in our code. @bartoszbetka ?

File: concent_api/api_testing_common.py (hunk 5 above)
Comment (Jakub89) [reply concent_api/api_testing_common.py]: Done

**Comment by rwrzesien** [comment]
@Jakub89 Then it was far beyond my jurisdiction ;)

**Comment by Jakub89** [comment]
@dybi I see right now that it can't be done, because tests for `not-found` and `bad-request` don't return `golem-messages-version` in header, so we can't give them `expected_golem_version` to put in header.
